proposals/interface-types/working-notes/Instructions.md

```diff
@@ -0,0 +1,277 @@
+
+This instruction has type `[sNN] -> [iMM]`
+
+The output is the value of `sNN` as a two's complement signed integer.
+
+If this cannot be represented in the desired type, then this instruction `fail`s.
+
+## `uNN.from_iMM`
+
+This instruction has type `iMM` -> `uNN`.
+
+The output is the value of `iMM` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction `fail`s.
+
+## `iMM.from_uNN`
+
+This instruction has type `[uNN] -> [iMM]`
+
+The output is the value of `uNN` as an unsigned integer.
+
+If this cannot be represented in the desired type, then this instruction `fail`s.
+
+## `string.size`
+
+The full syntax of this instruction is:
+```
+string.size #encoding
+```
+
+The type of this instruction is:
+```
+[string] -> [i32]
+```
+
+The output is the number of bytes needed to represent the input string in `#encoding`.
+
+The syntax for supported encodings is:
+```
+#encoding ::=
+  | utf8
+```
+
+Currently, only UTF8 is supported as an encoding. UTF16 may be added in the future.
+
+## `string.lift_memory`
+
+The full syntax of this instruction is:
+```
+string.lift_memory #memidx #encoding
+```
+
+The type of this instruction is:
+```
+[$base: i32, $len: i32] -> [string]
+```
+
+The output is a sequence of unicode code points taken by interpreting `store.memory[$memidx].bytes[$base .. $base + $len]` with `#encoding`.
+
+This instructions `fail`s if the range `$base .. $base + $len` is out-of-bounds, or the byte sequence is not valid under `#encoding`.
+
+## `string.lower_memory`
+
+The full syntax of this instruction is:
+```
+string.lower_memory #memidx #encoding
+```
+
+The type of this instruction is:
+```
+[$base: i32, string] -> []
+```
+
+The instruction will encode the string with `#encoding` into the memory specified by `#memidx` at the offset `$base`.
+
+## `record.lift`
+
+The full syntax of this instruction is:
+```
+record.lift #typeidx
+```
+
+The type of this instruction is:
+```
+[$a, $b, ..] -> [$type]
+
+Where:
+  * $type is a type of the form (record $fields) taken from #typeidx
+  * $a, $b, .. are the #interface-valtype's for each field in $fields
+```
+
+The output is a record interface value with fields given by the inputs.
+
+## `record.lower`
+
+The full syntax of this instruction is:
+```
+record.lower #typeidx
+```
+
+The type of this instruction is:
+```
+[$type] -> [$fields]
+
+Where:
+  * $type is a valid type of the form (record $fields) taken from #typeidx
+```
+
+The outputs are the fields of the input `record`.
+
+## `array.count`
+
+The full syntax of this instruction is:
+```
+array.count #typeidx
+```
+
+The type of this instruction is:
+```
+[$type] -> [i32]
+
+Where:
+  * $type is a type of the form (array ..) taken from #typeidx
+```
+
+The output of this instruction is the number of interface values contained in
+the input array.
+
+## `array.lift`
+
+This is a structured instruction with the following syntax:
+```
+array.lift #typeidx $stride: #s32
```

Comment:
I think this needs a #memidx and to be named `array.lift_memory` for the same reason as `string.lift_memory`, and similarly for lowering.

Reply:
The array instructions as written don't reference or require linear memory at all. `array.lift` loops over a variable starting at a `base`, `count` times, with an increment of `stride`. The inner block is responsible for using the variable to lift to the array's inner value. It may use the loop variable to index linear memory, but doesn't need to. `array.lower` is similar, except in the reverse direction.

The other way I was considering doing this was to somehow only have the segment of linear memory that is being lifted in an iteration of the loop available. Possibly by pushing a representation of it to the stack or as some new local-like thing? The goal would be to have something less expressive that could be optimized better, but I'm unsure how that should be represented.

Reply:
Hah, right you are; I had a stale old idea that the lifting and lowering were doing the writing into and out of linear memory.